api/v1alpha8/types.go

```diff
@@ -291,31 +291,61 @@ type LoadBalancer struct {
-	Rules []SecurityGroupRule `json:"rules"`
+type SecurityGroupStatus struct {
+	Name  string                    `json:"name"`
+	ID    string                    `json:"id"`
+	Rules []SecurityGroupRuleStatus `json:"rules"`
 }
 
-// SecurityGroupRule represent the basic information of the associated OpenStack
+// SecurityGroupRuleSpec represent the basic information of the associated OpenStack
 // Security Group Role.
-type SecurityGroupRule struct {
-	Description     string `json:"description"`
-	ID              string `json:"name"`
-	Direction       string `json:"direction"`
-	EtherType       string `json:"etherType"`
-	SecurityGroupID string `json:"securityGroupID"`
-	PortRangeMin    int    `json:"portRangeMin"`
-	PortRangeMax    int    `json:"portRangeMax"`
-	Protocol        string `json:"protocol"`
-	RemoteGroupID   string `json:"remoteGroupID"`
-	RemoteIPPrefix  string `json:"remoteIPPrefix"`
+// For now this is only used for the allNodesSecurityGroupRules but when we'll
+// other security groups, we'll need to add a validation because
+// Remote* fields are mutually exclusive.
+type SecurityGroupRuleSpec struct {
+	Description         string                     `json:"description,omitempty"`
+	Direction           string                     `json:"direction,omitempty"`
+	EtherType           string                     `json:"etherType,omitempty"`
+	PortRangeMin        int                        `json:"portRangeMin,omitempty"`
+	PortRangeMax        int                        `json:"portRangeMax,omitempty"`
+	Protocol            string                     `json:"protocol,omitempty"`
+	RemoteGroupID       string                     `json:"remoteGroupID,omitempty"`
+	RemoteIPPrefix      string                     `json:"remoteIPPrefix,omitempty"`
+	RemoteManagedGroups []ManagedSecurityGroupName `json:"remoteManagedGroups,omitempty"`
+}
+
+type SecurityGroupRuleStatus struct {
+	ID             string `json:"id"`
+	Description    string `json:"description,omitempty"`
+	Direction      string `json:"direction,omitempty"`
+	EtherType      string `json:"etherType,omitempty"`
+	PortRangeMin   int    `json:"portRangeMin,omitempty"`
+	PortRangeMax   int    `json:"portRangeMax,omitempty"`
+	Protocol       string `json:"protocol,omitempty"`
+	RemoteGroupID  string `json:"remoteGroupID,omitempty"`
+	RemoteIPPrefix string `json:"remoteIPPrefix,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=bastion;controlplane;worker
+type ManagedSecurityGroupName string
+
+func (m ManagedSecurityGroupName) String() string {
+	return string(m)
+}
+
+// SecurityGroupsSpec defines the desired state of security groups and rules for the cluster.
+type SecurityGroupsSpec struct {
+	// Enabled defines whether security groups should be created.
+	Enabled bool `json:"enabled,omitempty"`
+
+	// AllNodesSecurityGroupRules defines the rules that should be applied to all nodes.
+	AllNodesSecurityGroupRules []SecurityGroupRuleSpec `json:"allNodesSecurityGroupRules,omitempty"`
```

Comment:
yeah I was thinking this as well but I couldn't think of a key too. I'm really not sure about `name` :thinking: - maybe @mdbooth has some idea?

Reply:
It's a tricky one. I agree that users are going to want to merge this. Adding a 'fake' name to use as a merge key, though... I guess it's not completely gross.

@JoelSpeed are you aware of any prior art?

Reply:
Is there a field on openstack for this security group rule that must be unique? Does the description have to be unique? Or maybe the combination of source/destination/port/protocol?

You can use more than key for your `listMapKey`

Example of expected keys https://github.com/openshift/api/blob/4f00b4f16b634830bb16b432eec91f6e514f2981/machine/v1/types_controlplanemachineset.go#L403-L408

Reply:
> Is there a field on openstack for this security group rule that must be unique? Does the description have to be unique? Or maybe the combination of source/destination/port/protocol?

There's no single unique field which is actually used by OpenStack. Using a combination of fields is an interesting idea. Unfortunately I think the set of key fields would have to be all fields. That could work, but it would mean that adding a field to the struct would become a breaking change.

Summarising:

Adding `name` requires the user to create a new field which is only used by the CAPO API, but can be easily added/updated/deleted by merge.

Using all struct fields doesn't require an extraneous field. Users can easily add new rules by merging. Deleting requires specifying the whole rule. Updating isn't possible directly, but can be achieved with delete and add. The API is no longer extensible without updating the merge key, which I think is a breaking change (e.g. it would break existing delete 

When it comes to merging, adding a fake merge key certainly feels like the better option. Is it worth requiring users to always create a name field?

Reply:
> When it comes to merging, adding a fake merge key certainly feels like the better option. Is it worth requiring users to always create a name field?

You wouldn't be the first to do so. Though it's a shame there's no small combination of fields that can identify a rule as unique

Reply:
@EmilienM I think the resolution here is that we'll add a fake, required name attribute and make it the merge key.

Reply:
+1 for that :slightly_smiling_face:

Reply:
sounds good; thanks @JoelSpeed for your guidance, super appreciated as always :-)

Reply:
I've addressed this feedback on a separated commit, but I'll squash once this PR is approved.
Note: I've added webhook validations for various fields. I know this has caused some concerns before, so let me know in the review if you want me to drop them.

Reply:
@mdbooth FYI2 also I've removed `omitempty` and added docstrings + some kubebuilder.